contracts/VeJoeStaking.sol

```diff
@@ -0,0 +1,267 @@
+        );
+        boostedGenerationRate = _boostedGenerationRate;
+    }
+
+    /// @notice Set boostedThreshold
+    /// @param _boostedThreshold the new boostedThreshold
+    function setBoostedThreshold(uint256 _boostedThreshold) external onlyOwner {
+        require(
+            _boostedThreshold <= 100,
+            "VeJoeStaking: expected new _boostedThreshold to be less than or equal to 100"
+        );
+        boostedThreshold = _boostedThreshold;
+    }
+
+    /// @notice Set boostedDuration
+    /// @param _boostedDuration the new boostedDuration
+    function setBoostedDuration(uint256 _boostedDuration) external onlyOwner {
+        boostedDuration = _boostedDuration;
+    }
+
+    /// @notice Deposits JOE to start staking for veJOE. Note that any pending veJOE
+    /// will also be claimed in the process.
+    /// @param _amount the amount of JOE to deposit
+    function deposit(uint256 _amount) external {
+        require(_amount > 0, "VeJoeStaking: expected deposit amount to be greater than zero");
+
+        if (getUserHasNonZeroBalance(msg.sender)) {
+            // If user already has staked JOE, we first send them any pending veJOE
+            _claim();
+
+            userInfos[msg.sender].balance += _amount;
+
+            // User is eligible for boosted benefits if:
+            // 1. They are not already currently receiving boosted benefits
+            // 2. Their staked JOE is at least `boostedThreshold / 100 * totalStakedJoe`
+            if (userInfos[msg.sender].boostEndTimestamp == 0) {
+                uint256 totalStakedJoe = joe.balanceOf(address(this));
+                if ((userInfos[msg.sender].balance * 100) / boostedThreshold >= totalStakedJoe) {
+                    userInfos[msg.sender].boostEndTimestamp = block.timestamp + boostedDuration;
+                }
+            }
+        } else {
+            // If the user's `lastRewardTimestamp` is 0, i.e. if this is the user's first time staking,
+            // then they will receive boosted benefits.
+            // Note that it is important we perform this check **before** we update the user's `lastRewardTimestamp`
+            // down below.
+            if (userInfos[msg.sender].lastRewardTimestamp == 0) {
+                userInfos[msg.sender].boostEndTimestamp = block.timestamp + boostedDuration;
+            }
+            userInfos[msg.sender].balance = _amount;
+            userInfos[msg.sender].lastRewardTimestamp = block.timestamp;
+        }
+
+        joe.safeTransferFrom(msg.sender, address(this), _amount);
+
+        emit Deposit(msg.sender, _amount);
+    }
+
+    /// @notice Withdraw staked JOE. Note that unstaking any amount of JOE means you will
+    /// lose all of your current veJOE.
+    /// @param _amount the amount of JOE to unstake
+    function withdraw(uint256 _amount) external {
+        require(_amount > 0, "VeJoeStaking: expected to withdraw non-zero amount of JOE");
```

Comment:
Should we be recalculating the users `boostEndTimestamp` here? Right now I could deposit a large amount of JOE to achieve a boosted generation rate and then withdraw it instantly (this flashloan).

Reply:
I don't believe so because anytime you withdraw `JOE`, all of your `veJOE` will be burned